espnet/nets/chainer_backend/rnn/training.py

```diff
@@ -73,6 +77,8 @@ def __init__(self, train_iter, optimizer, converter, device, accum_grad=1):
         self.forward_count = 0
         self.accum_grad = accum_grad
         self.start = True
+        # To solve #1091, it is required to set the variable inside this class.
+        self.device = device
```

Comment:
I do not think you need this because base class does
https://github.com/chainer/chainer/blob/a3c428a32905052e33fc6427f5bfd8ae4b782f79/chainer/training/updaters/standard_updater.py#L92

Reply:
Problem maybe be caused from this line.
https://github.com/espnet/espnet/blob/63b9cb451ab17e17506347f29cd0837d978647a8/espnet/nets/chainer_backend/rnn/training.py#L234
In base class, `self.device` is converted to `CpuDevice`.

https://github.com/chainer/chainer/blob/a3c428a32905052e33fc6427f5bfd8ae4b782f79/chainer/training/updaters/standard_updater.py#L71

Therefore, we should change if-condition like:
```python
xp = cuda.cupy if not isinstance(device, chainer.backends.CpuDevice) else np 
```

Reply:
The interesting part is that device is `chainer.backends.CpuDevice` in the first iter only, after that it becomes a int (-1 or the gpuid). I was also expecting to be a `CpuDevice` the whole training so it will be possible to extract the xp from the self device (So the if is no longer requires) but sadly it was not as expected. This is the only solution which I found without adding more lines. If you tried any other besides the mentioned let me know, but I could not find it.. T.T

Reply:
Now I understood your point.
I doubt that following part:
https://github.com/espnet/espnet/blob/63b9cb451ab17e17506347f29cd0837d978647a8/espnet/asr/chainer_backend/asr.py#L286-L289
https://github.com/espnet/espnet/blob/63b9cb451ab17e17506347f29cd0837d978647a8/espnet/asr/asr_utils.py#L80
But easiest solution seems yours. 
Why don't your add `TODO(...):` comment in this time?
@ShigekiKarita how do you think?